models/classes/runner/time/QtiTimer.php

```diff
@@ -0,0 +1,340 @@
+        return $duration >= $timeLimit;
+    }
+
+    /**
+     * Sets the storage used to maintain the data
+     * @param TimeStorage $storage
+     * @return Timer
+     */
+    public function setStorage(TimeStorage $storage)
+    {
+        $this->storage = $storage;
+        return $this;
+    }
+
+    /**
+     * Gets the storage used to maintain the data
+     * @return TimeStorage
+     */
+    public function getStorage()
+    {
+        return $this->storage;
+    }
+
+    /**
+     * Saves the data to the storage
+     * @return Timer
+     * @throws InvalidStorageException
+     * @throws InvalidDataException
+     * @throws \common_exception_Error
+     */
+    public function save()
+    {
+        if (!$this->storage) {
+            throw new InvalidStorageException('A storage must be defined in order to store the data!');
+        }
+        
+        $this->storage->store(serialize($this->timeLine));
+        
+        return $this;
+    }
+
+    /**
+     * Loads the data from the storage
+     * @return Timer
+     * @throws InvalidStorageException
+     * @throws InvalidDataException
+     * @throws \common_exception_Error
+     */
+    public function load()
+    {
+        if (!$this->storage) {
+            throw new InvalidStorageException('A storage must be defined in order to store the data!');
+        }
+        
+        $data = $this->storage->load();
+        
+        if (isset($data)) {
+            $this->timeLine = unserialize($data);
+
+            if (!$this->timeLine instanceof TimeLine) {
+                throw new InvalidDataException('The storage did not provide acceptable data when loading!');
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Checks if a timestamp is consistent with existing TimePoint within a range
+     * @param array $points
+     * @param float $timestamp
+     * @throws InconsistentRangeException
+     */
+    protected function checkTimestampCoherence($points, $timestamp)
+    {
+        foreach($points as $point) {
+            if ($point->getTimestamp() > $timestamp) {
+                throw new InconsistentRangeException('A new TimePoint cannot be set before an existing one!');
+            }
+        }
+    }
+    
+    /**
+     * Gets the tags describing a particular item with an assessment test
+     * @param RouteItem $routeItem
+     * @return array
+     */
+    protected function getItemTags(RouteItem $routeItem)
+    {
+        $tags = [];
+
+        $itemRef = $routeItem->getAssessmentItemRef();
+        $testPart = $routeItem->getTestPart();
+        $sections = $routeItem->getAssessmentSections();
+
+        $occurrence = $routeItem->getOccurence();
+        $partId = $testPart->getIdentifier();
+        $sectionId = key(current($sections));
+        $itemId = $itemRef->getIdentifier();
+        
+        $tags []= $itemId;
+        $tags []= $itemId . '#' . $occurrence;
+        $tags []= $partId;
+        $tags []= $sectionId;
```

Comment:
A time line is related to a test, so if we compute the duration without any criteria, the result will embrace the whole test.